src/crimson/osd/backfill_state.cc

```diff
@@ -0,0 +1,448 @@
+               bs().last_backfill_started));
+  }
+  bs().backfill_info.trim_to(bs().last_backfill_started);
+}
+
+/* static */ bool BackfillState::Enqueuing::all_peer_enqueued(
+  const PeeringFacade& ps,
+  const std::map<pg_shard_t, BackfillInterval>& peer_backfill_info)
+{
+  for (const auto& bt : ps.get_backfill_targets()) {
+    const auto piter = peer_backfill_info.find(bt);
+    ceph_assert(piter != peer_backfill_info.end());
+    const BackfillInterval& pbi = piter->second;
+    // See if peer has more to process
+    if (!pbi.extends_to_end() || !pbi.empty()) {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool BackfillState::Enqueuing::all_peer_enqueued(
+  const std::map<pg_shard_t, BackfillInterval>& peer_backfill_info) const
+{
+  // Primary hasn't got any more objects
+  ceph_assert(bs().backfill_info.empty());
+  return all_peer_enqueued(ps(), peer_backfill_info);
+}
+
+/* static */ bool BackfillState::Enqueuing::all_enqueued(
+  const PeeringFacade& ps,
+  const BackfillInterval& backfill_info,
+  const std::map<pg_shard_t, BackfillInterval>& peer_backfill_info)
+{
+  const bool all_local_enqueued = \
+    backfill_info.extends_to_end() && backfill_info.empty();
+  const bool all_peer_enqueued =  \
+    Enqueuing::all_peer_enqueued(ps, peer_backfill_info);
+  return all_local_enqueued && all_peer_enqueued;
+}
+
+hobject_t BackfillState::Enqueuing::earliest_peer_backfill(
+  const std::map<pg_shard_t, BackfillInterval>& peer_backfill_info) const
+{
+  hobject_t e = hobject_t::get_max();
+  for (const pg_shard_t& bt : ps().get_backfill_targets()) {
+    const auto iter = peer_backfill_info.find(bt);
+    ceph_assert(iter != peer_backfill_info.end());
+    e = std::min(e, iter->second.begin);
+  }
+  return e;
+}
+
+bool BackfillState::Enqueuing::should_rescan_replicas(
+  const std::map<pg_shard_t, BackfillInterval>& peer_backfill_info,
+  const BackfillInterval& backfill_info) const
+{
+  const auto& targets = ps().get_backfill_targets();
+  return std::any_of(std::begin(targets), std::end(targets),
+    [&, this] (const auto& bt) {
+      return ReplicasScanning::replica_needs_scan(peer_backfill_info.at(bt),
+                                                  backfill_info);
+    });
+}
+
+bool BackfillState::Enqueuing::should_rescan_primary(
+  const std::map<pg_shard_t, BackfillInterval>& peer_backfill_info,
+  const BackfillInterval& backfill_info) const
+{
+  return backfill_info.begin <= earliest_peer_backfill(peer_backfill_info) &&
+	 !backfill_info.extends_to_end();
+}
+
+void BackfillState::Enqueuing::trim_backfilled_object_from_intervals(
+  BackfillState::Enqueuing::result_t&& result,
+  hobject_t& last_backfill_started,
+  std::map<pg_shard_t, BackfillInterval>& peer_backfill_info)
+{
+  std::for_each(std::begin(result.pbi_targets), std::end(result.pbi_targets),
+    [this, &peer_backfill_info] (const auto& bt) {
+      peer_backfill_info.at(bt).pop_front();
```

Comment:
I think `at()` was chosen because it offers:
* more concise code than the `find()`-based counterpart (assumed here that shorter code is easier to read),
* the `const` restriction. In contrast to `operator[]`, `at()` has a `const`-qualified variant.